FSI/Force.C

```diff
@@ -28,7 +28,53 @@ void preciceAdapter::FSI::Force::write(double* buffer, bool meshConnectivity, co
 
 void preciceAdapter::FSI::Force::read(double* buffer, const unsigned int dim)
 {
-    this->readFromBuffer(buffer);
+    // Copy the force field from the buffer to OpenFOAM
+
+    // Here we assume that a force volVectorField exists, which is used by
+    // the OpenFOAM solver
+
+    // Lookup the force field name
+    const word forceFieldName
+    (
+        mesh_.lookupObject<IOdictionary>
+        (
+            "preciceDict"
+        ).subDict("FSI").lookup("forceFieldName")
+    );
```

Comment:
This is probably not in keeping with the preCICE coding style. This was the only way I could figure out how to access the preciceDict from the Force class. I am open to suggestions on a nicer way to access this.

Reply:
We currently store similar fields in the `FluidStructureInteraction` class: https://github.com/precice/openfoam-adapter/blob/develop/FSI/FSI.H#L31-L35

We then read them in the `readConfig` method: https://github.com/precice/openfoam-adapter/blob/68d782bd506273c4c5ce2f48cad13fcfcf899a41/FSI/FSI.C#L48

Reply:
I have now added a solidForceName to the FSI and Force classes in 75eca1b.

In the "void preciceAdapter::FSI::Force::read(double* buffer, const unsigned int dim)" function, I still need to look up the force field (as opposed to storing a reference to it in the Force class) as
```
    volVectorField& forceField =
        const_cast<volVectorField&>(
            mesh_.lookupObject<volVectorField>(nameSolidForce_));
```
because when the OpenFOAM fluid solvers create an object of type "Force" then they would not have a "solidForce" field.

An alternative would be to add a volVectorField pointer to the solidForce field and allow the pointer to be NULL for fluid solvers; in that case, care would be needed when using the pointer.

Reply:
This is a bit inconsistent with how we do things in the rest of the adapter (which I know can be improved, but we should do that in a separate step later).

See, for example, the `Displacement.[H|C]`. In the constructor, we get a pointer to the field, which we store in the class. We then just manipulate this pointer in the read/write. Why is this approach not fitting here?

You can also store the name in the class (initialized in the constructor). I know it is currently very verbose and a bit ugly how we just forward arguments, but we can fix this separately in a larger refactoring.

By the way, I think we should actually modify the `ForceBase.C` file instead of `Force.C` (@DavidSCN please comment).

Reply:
I did not store a pointer to the field because this pointer will be null when the adapter is used for fluids; however, since we are already using a nullptr in the Displacement class, I can do the same thing here. See my latest changes in e18ea6e and 3b17ab0.

Depending on what you and @DavidSCN decide regarding using nullptrs, the same approach can be adopted here.

One alternative option would be to use the lazy evaluation approach which is popular in OpenFOAM (may be overkill here), i.e.

- store pointers to the fields (as you do now), but initialise them to nullptr by default
- add an access function for each field, which checks if the pointer is assigned and then returns the field; if it is unassigned then it looks it up or returns an error if not found.
- In that way, developers should avoid the direct use of the pointers and instead access the fields via the access functions.

Reply:
For FSI simulations, the adapter supports writing `Force` and `Stress` data and computing these two quantities is pretty much the same, the computation only differs by the face area multiplication. For this reason, we created a common base class having common functions such as `writeToBuffer`.
When reading data, the situation is a bit different as we don't compute anything in the adapter itself, but just 'apply' the values. Hence, the functionality is for `Force` and `Stress` data the same, but it's also the same for other coupling data. As already mentioned by @MakisH, things are not optimal in the adapter and that's actually one of the things duplicated a lot in the code.
I would only move this to the base class if solids4foam supports reading `Stress` data, because I'm not aware of any other OpenFOAM solver who can handle this. If we face demand for reading `Stress` data, we can easily move it (or rather refactor the 'reading' infrastructure at some point).

Reply:
Regarding the `nullptr`, I would vote for the access method, as it is a bit safer, but I did not follow the discussion as closely as you did, so feel free to implement what you think is the most fitting approach.

Reply:
I understand that this is fine now, and we can postpone any refactoring to the moment where this will actually be needed.